src/Agent.Listener/SelfUpdater.cs

```diff
@@ -155,6 +168,45 @@ private async Task<bool> UpdateNeeded(string targetVersion, CancellationToken to
             return false;
         }
 
+        private async Task<bool> HashValidation(string archiveFile)
+        {
+            if (_hashValidationDisabled)
+            {
+                Trace.Info($"Agent package hash validation disabled, so skipping it");
+                return true;
+            }
+
+            bool isHostedServer = await _serverUtil.IsDeploymentTypeHosted(_serverUrl, _creds, _locationServer);
+
+            if (!isHostedServer)
+            {
+                Trace.Info($"Skipping checksum validation for On-Premises solution");
+                return true;
+            }
+
+            if (string.IsNullOrEmpty(_targetPackage.HashValue))
+            {
+                Trace.Warning($"Unable to perform the necessary checksum validation since the target package hash is missed");
+                return false;
+            }
+
+            string archiveHash = IOUtil.GetFileHash(archiveFile);
+            bool hashesMatch = StringUtil.HashesMatch(archiveHash, _targetPackage.HashValue);
+
+            if (hashesMatch)
+            {
+                Trace.Info($"Checksum validation succeeded");
+                return true;
+            }
+
+            // A hash mismatch can occur in two cases:
+            // 1) The archive is compromised
+            // 2) The archive was not fully downloaded or was damaged during downloading
+            // There is no way to determine the case so we just return false in both cases (without throwing an exception)
+            Trace.Warning($"Checksum validation failed");
```

Comment:
Could it make sense to also display the actual and expected hash sum in the error message? I think it may help with the investigation if something goes wrong in the validation process.